File: IOS/Claude45/appomodoro/appomodoro/Utils/AngleMath.swift
//
//  AngleMath.swift
//  appomodoro
//
//  Created by JOSE ZARABANDA on 10/8/25.
//

import Foundation
import CoreGraphics

/// Utility functions for angle calculations in the dial view
struct AngleMath {
    
    /// Calculate the rotation angle for seconds ring
    /// - Parameter elapsed: elapsed time in seconds
    /// - Returns: angle in radians (starts at top, -π/2, rotates clockwise)
    static func secondsAngle(elapsed: TimeInterval) -> Double {
        let secondsFrac = elapsed.truncatingRemainder(dividingBy: 60)  // 0..<60
        return -Double.pi / 2 + 2 * Double.pi * (secondsFrac / 60.0)   // 1 full turn per 60s
    }
    
    /// Calculate the rotation angle for minutes ring
    /// - Parameter elapsed: elapsed time in seconds
    /// - Returns: angle in radians (starts at top, -π/2, rotates clockwise)
    static func minutesAngle(elapsed: TimeInterval) -> Double {
        let minutesFrac = (elapsed / 60.0).truncatingRemainder(dividingBy: 60)  // 0..<60 minutes
        return -Double.pi / 2 + 2 * Double.pi * (minutesFrac / 60.0)            // 1 full turn per 3600s
    }
    
    /// Get position on circle for given angle and radius
    /// - Parameters:
    ///   - angle: angle in radians
    ///   - radius: radius of the circle
    ///   - center: center point of the circle
    /// - Returns: CGPoint on the circumference
    static func pointOnCircle(angle: Double, radius: CGFloat, center: CGPoint) -> CGPoint {
        let x = center.x + radius * CGFloat(cos(angle))
        let y = center.y + radius * CGFloat(sin(angle))
        return CGPoint(x: x, y: y)
    }
    
    /// Format time interval as MM:SS
    /// - Parameter interval: time in seconds
    /// - Returns: formatted string "MM:SS"
    static func formatTime(_ interval: TimeInterval) -> (minutes: String, seconds: String) {
        let totalSeconds = max(0, Int(interval.rounded()))
        let minutes = totalSeconds / 60
        let seconds = totalSeconds % 60
        return (String(format: "%02d", minutes), String(format: "%02d", seconds))
    }
    
    /// Get the current value that should appear in selector window
    /// - Parameters:
    ///   - elapsed: elapsed time
    ///   - isSeconds: true for seconds ring, false for minutes ring
    /// - Returns: the number (0-59) that appears in the selector
    static func currentDisplayValue(elapsed: TimeInterval, isSeconds: Bool) -> Int {
        if isSeconds {
            return Int(elapsed.truncatingRemainder(dividingBy: 60))
        } else {
            return Int((elapsed / 60.0).truncatingRemainder(dividingBy: 60))
        }
    }
    
    /// Normalize angle to 0...2π range
    static func normalizeAngle(_ angle: Double) -> Double {
        var normalized = angle.truncatingRemainder(dividingBy: 2 * .pi)
        if normalized < 0 {
            normalized += 2 * .pi
        }
        return normalized
    }
}

File: Android/Claude_45/app/src/main/java/com/zarabandajose/appomodoro/util/AngleMath.kt
package com.zarabandajose.appomodoro.util

import kotlin.math.PI

object AngleMath {
    /**
     * Calculate the angle for the seconds ring rotation.
     * Formula: θ = -π/2 + 2π * ((elapsedSec % 60) / 60)
     * 
     * @param elapsedSeconds Total elapsed seconds (can be > 60)
     * @return Angle in radians for seconds ring rotation
     */
    fun calculateSecondsAngle(elapsedSeconds: Double): Float {
        val normalizedSeconds = elapsedSeconds % 60.0
        return (-PI / 2.0 + 2.0 * PI * (normalizedSeconds / 60.0)).toFloat()
    }
    
    /**
     * Calculate the angle for the minutes ring rotation.
     * Formula: θ = -π/2 + 2π * ((elapsedSec / 60) / 60)
     * This creates a smooth minutes rotation (1 full turn per 3600 seconds = 1 hour)
     * 
     * @param elapsedSeconds Total elapsed seconds
     * @return Angle in radians for minutes ring rotation
     */
    fun calculateMinutesAngle(elapsedSeconds: Double): Float {
        val minutesFraction = (elapsedSeconds / 60.0) / 60.0
        return (-PI / 2.0 + 2.0 * PI * minutesFraction).toFloat()
    }
    
    /**
     * Calculate position on a circle given angle and radius.
     * Returns x offset from center.
     */
    fun getXOnCircle(angle: Float, radius: Float): Float {
        return radius * kotlin.math.cos(angle)
    }
    
    /**
     * Calculate position on a circle given angle and radius.
     * Returns y offset from center.
     */
    fun getYOnCircle(angle: Float, radius: Float): Float {
        return radius * kotlin.math.sin(angle)
    }
}
